#! /bin/sh /usr/share/dpatch/dpatch-run
## 2683_possible_crash_update_speaker_list.patch.dpatch by  <root@debian.example.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: 2683_possible_crash_update_speaker_list.patch

@DPATCH@

--- linden/indra/newview/llfloateractivespeakers.cpp	2007-09-20 12:36:50.000000000 +0200
+++ linden/indra/newview/llfloateractivespeakers.cpp	2007-10-01 16:39:42.562500000 +0200
@@ -806,7 +903,7 @@
 	for(avatar_it = LLCharacter::sInstances.begin(); avatar_it != LLCharacter::sInstances.end(); ++avatar_it)
 	{
 		LLVOAvatar* avatarp = (LLVOAvatar*)*avatar_it;
-		if (dist_vec(avatarp->getPositionAgent(), gAgent.getPositionAgent()) <= CHAT_NORMAL_RADIUS)
+		if (!avatarp->isDead() &&  dist_vec(avatarp->getPositionAgent(), gAgent.getPositionAgent()) <= CHAT_NORMAL_RADIUS)
 		{
 			setSpeaker(avatarp->getID());
 		}
@@ -820,7 +917,7 @@
 		if (speakerp->mStatus == LLSpeaker::STATUS_TEXT_ONLY)
 		{
 			LLVOAvatar* avatarp = (LLVOAvatar*)gObjectList.findObject(speaker_id);
-			if (!avatarp || dist_vec(avatarp->getPositionAgent(), gAgent.getPositionAgent()) > CHAT_NORMAL_RADIUS)
+			if (!avatarp || avatarp->isDead() || dist_vec(avatarp->getPositionAgent(), gAgent.getPositionAgent()) > CHAT_NORMAL_RADIUS)
 			{
 				speakerp->mStatus = LLSpeaker::STATUS_NOT_IN_CHANNEL;
 				speakerp->mDotColor = INACTIVE_COLOR;
--- linden/indra/newview/llviewerobjectlist.cpp	2007-09-20 12:36:50.000000000 +0200
+++ linden/indra/newview/llviewerobjectlist.cpp	2007-10-01 16:45:38.562500000 +0200
@@ -868,6 +868,10 @@
 		if (objectp->mRegionp == regionp)
 		{
 			killObject(objectp);
+
+			// invalidate region pointer. region will become invalid, but 
+			// refcounted objects may survive the cleanDeadObjects() call below
+			objectp->mRegionp = NULL;	 
 		}
 	}
 
